### PR TITLE
ARROW-16918: [Gandiva][C++] Adding UTC-local timezone conversion functions

### PR DESCRIPTION
**Pull request by palak-9202**
Adding functions 
to_utc_timezone : Converts a timestamp from local timezone to UTC time
from_utc_timezone : Converts a timestamp from UTC time to local time

### CODE REVIEW

File: cpp/src/gandiva/gdv_function_stubs.cc
Comment (anthonylouisbsb) [inline cpp/src/gandiva/gdv_function_stubs.cc:628]: As it is an implementation of a Hive function, do you know if there is a way to print all available timezones for the arrow_vendored library? Is just to check the compatibility of that function with all available timezones of Hive: http://obscuredclarity.blogspot.com/2010/08/get-all-timezones-available-in-timezone.html

**Review by anthonylouisbsb — CHANGES_REQUESTED** [review]
I have only one question about the library you used. Also, check the Travis test that is failling.

File: cpp/src/gandiva/gdv_function_stubs.cc
Comment (palak-9202) [reply cpp/src/gandiva/gdv_function_stubs.cc:628]: Try:
for(auto const& z : get_tzdb().zones) {
    cout<<z.name()<<'\n';
  }

**Review by anthonylouisbsb — APPROVED** [review]

File: cpp/src/gandiva/function_registry_datetime.cc
Comment (kou) [inline cpp/src/gandiva/function_registry_datetime.cc:181]: Could you revert this change before we merge this?

File: cpp/src/gandiva/gdv_function_stubs.cc
Comment (kou) [inline cpp/src/gandiva/gdv_function_stubs.cc:631]:
```suggestion
    gdv_fn_context_set_error_msg(context, "Invalid time zone");
    return 0;
```

File: cpp/src/gandiva/gdv_function_stubs.cc
Comment (kou) [inline cpp/src/gandiva/gdv_function_stubs.cc:626]:
```suggestion
    const auto local_tz = locate_zone(std::string(timezone, length));
```

File: cpp/src/gandiva/gdv_function_stubs.cc
Comment (kou) [inline cpp/src/gandiva/gdv_function_stubs.cc:646]: Can we use `const auto utc_tz = ...` here?

File: cpp/src/gandiva/gdv_function_stubs.cc
Comment (kou) [inline cpp/src/gandiva/gdv_function_stubs.cc:654]: How about adding the given `timezone` to error message? It will help users to fix the invalid input.

File: cpp/src/gandiva/precompiled/time.cc
Comment (kou) [inline cpp/src/gandiva/precompiled/time.cc:1058]:
```suggestion
```

File: cpp/src/gandiva/precompiled/types.h
Comment (kou) [inline cpp/src/gandiva/precompiled/types.h:436]:
```suggestion
```

File: cpp/src/gandiva/gdv_function_stubs_test.cc
Comment (kou) [inline cpp/src/gandiva/gdv_function_stubs_test.cc:1022]: How about adding a test for failure case too?

**Comment by kou** [comment]
We can fix style by `cmake --build ${BUILD_DIR} --target format` but you need to install `clang-format-12`.
If it's difficult to prepare `clang-format-12`, I can fix style and push the fix to this branch.

File: cpp/src/gandiva/gdv_function_stubs.cc
Comment (palak-9202) [reply cpp/src/gandiva/gdv_function_stubs.cc:646]: I had used 'auto' earlier but this caused a build error because this object is passed to an overloaded function, so the type had to be explicitly specified

File: cpp/src/gandiva/gdv_function_stubs.cc
Comment (palak-9202) [reply cpp/src/gandiva/gdv_function_stubs.cc:654]: Sure

File: cpp/src/gandiva/gdv_function_stubs_test.cc
Comment (palak-9202) [reply cpp/src/gandiva/gdv_function_stubs_test.cc:1022]: I've added a failure test and modified the error message to include the time zone name given as input

File: cpp/src/gandiva/gdv_function_stubs.cc
Comment (kou) [reply cpp/src/gandiva/gdv_function_stubs.cc:654]: Umm, I couldn't reproduce it with

```diff
diff --git a/cpp/src/gandiva/gdv_function_stubs.cc b/cpp/src/gandiva/gdv_function_stubs.cc
index ccf3e7ca3b..e6cbeb8aa7 100644
--- a/cpp/src/gandiva/gdv_function_stubs.cc
+++ b/cpp/src/gandiva/gdv_function_stubs.cc
@@ -651,11 +651,9 @@ gdv_timestamp from_utc_timezone_timestamp(gdv_int64 context,
   using arrow_vendored::date::make_zoned;
 
   sys_time <milliseconds> tp {milliseconds{time_miliseconds}};
-  const zoned_time<milliseconds, const time_zone*> utc_tz =
-                      make_zoned(std::string("Etc/UTC"), tp);
+  const auto utc_tz = make_zoned(std::string("Etc/UTC"), tp);
   try {
-    const zoned_time<milliseconds, const time_zone*> local_tz =
-                      make_zoned(std::string(timezone, length), utc_tz);
+    const auto local_tz = make_zoned(std::string(timezone, length), utc_tz);
     gdv_timestamp offset = local_tz.get_time_zone()->get_info(tp).offset.count()*1000;
     return time_miliseconds + static_cast<gdv_timestamp>(offset);
   } catch(...) {
```

Could you show an error message on your environment?

File: cpp/src/gandiva/gdv_function_stubs_test.cc
Comment (kou) [inline cpp/src/gandiva/gdv_function_stubs_test.cc:1002]: Is this correct?

File: cpp/src/gandiva/gdv_function_stubs_test.cc
Comment (kou) [inline cpp/src/gandiva/gdv_function_stubs_test.cc:1009]: Is this correct?
It seems that `18000000` is "1970-01-01 5:00:00 UTC" and "1970-01-01 5:00:00 Asia/Kolkata" is `-1800000`.

File: cpp/src/gandiva/gdv_function_stubs.cc
Comment (kou) [inline cpp/src/gandiva/gdv_function_stubs.cc:638]: @anthonylouisbsb Could you confirm whether this is a preferred error message generation approach in Gandiva? (I'm not familiar with Gandiva's error message generation.)

File: cpp/src/gandiva/gdv_function_stubs_test.cc
Comment (palak-9202) [reply cpp/src/gandiva/gdv_function_stubs_test.cc:1009]: You're right, I'll make the changes for both these cases

File: cpp/src/gandiva/gdv_function_stubs.cc
Comment (palak-9202) [reply cpp/src/gandiva/gdv_function_stubs.cc:654]: I didn't get any errors on my environment either, the build error occurred in the Appveyor build check

Comment (kou) [reply cpp/src/gandiva/gdv_function_stubs.cc:654]: OK. Is this the error log? https://ci.appveyor.com/project/ApacheSoftwareFoundation/arrow/builds/44007267/job/2upjodye0ip00r66
And is this the fix of the error? https://github.com/palak-9202/arrow/commit/4c0549526e56866205e4b1baa993b2e3e67241d5

Comment (palak-9202) [reply cpp/src/gandiva/gdv_function_stubs.cc:654]: Yes

File: cpp/src/gandiva/gdv_function_stubs_test.cc
Comment (palak-9202) [reply cpp/src/gandiva/gdv_function_stubs_test.cc:1009]: Hey, actually since the function takes in the timestamp separate from the time zone, the time stamp is not adjusted for zone offset (this is done to match the hive function implementation). so what the function receives is a timestamp that should be interpreted as the time in the zone given as second parameter. That's why the function subtracts the offset to get UTC time even though the actual equivalent timestamp is the same, just the representation of the date+time is different.
So my comments are misleading- the input is just a date + time whose timestamp will be casted based on 0 offset. See the projector test about this.
I will remove the time zone names from the comments.

File: cpp/src/gandiva/gdv_function_stubs.cc
Comment (kou) [reply cpp/src/gandiva/gdv_function_stubs.cc:654]: It seems that https://github.com/palak-9202/arrow/commit/4c0549526e56866205e4b1baa993b2e3e67241d5#diff-6112a00ef24accb25d6dde3f15276ee354c695838994d2e80e0de65c0e1929abL100 is the problem. https://github.com/palak-9202/arrow/commit/4c0549526e56866205e4b1baa993b2e3e67241d5#diff-6112a00ef24accb25d6dde3f15276ee354c695838994d2e80e0de65c0e1929abL102 didn't cause the problem.
The former uses `const char *` and the latter uses `std::string`.
How about trying the above diff that uses `std::string` for the former?

Comment (palak-9202) [reply cpp/src/gandiva/gdv_function_stubs.cc:654]: I'll try it out

Comment (palak-9202) [reply cpp/src/gandiva/gdv_function_stubs.cc]: It worked out 😸

File: cpp/src/gandiva/gdv_function_stubs.cc
Comment (kou) [inline cpp/src/gandiva/gdv_function_stubs.cc:621]:
```suggestion
```

File: cpp/src/gandiva/gdv_function_stubs.cc
Comment (kou) [inline cpp/src/gandiva/gdv_function_stubs.cc:650]:
```suggestion
```

File: cpp/src/gandiva/gdv_function_stubs.cc
Comment (kou) [inline cpp/src/gandiva/gdv_function_stubs.cc:631]:
```suggestion
    auto err_msg =
```

File: cpp/src/gandiva/gdv_function_stubs.cc
Comment (kou) [inline cpp/src/gandiva/gdv_function_stubs.cc:630]:
```suggestion
    auto msg_len = static_cast<int32_t>(strlen(timezone) + 50);
```

File: cpp/src/gandiva/gdv_function_stubs.cc
Comment (kou) [inline cpp/src/gandiva/gdv_function_stubs.cc:660]:
```suggestion
    auto msg_len = static_cast<int32_t>(strlen(timezone) + 50);
```

File: cpp/src/gandiva/gdv_function_stubs.cc
Comment (kou) [inline cpp/src/gandiva/gdv_function_stubs.cc:661]:
```suggestion
    auto err_msg =
```

File: cpp/src/gandiva/gdv_function_stubs_test.cc
Comment (kou) [inline cpp/src/gandiva/gdv_function_stubs_test.cc:1000]:
```suggestion
  auto len_ist = static_cast<gdv_int32>(strlen("Asia/Kolkata"));
  auto len_pst = static_cast<gdv_int32>(strlen("America/Los_Angeles"));
```

File: cpp/src/gandiva/gdv_function_stubs_test.cc
Comment (kou) [inline cpp/src/gandiva/gdv_function_stubs_test.cc:1015]:
```suggestion
  ts = 1520730000000;
```

File: cpp/src/gandiva/gdv_function_stubs_test.cc
Comment (kou) [inline cpp/src/gandiva/gdv_function_stubs_test.cc:1020]: Is this correct?
It seems that 1331712000000 is "2012-03-14 08:00:00 UTC".

File: cpp/src/gandiva/gdv_function_stubs_test.cc
Comment (kou) [inline cpp/src/gandiva/gdv_function_stubs_test.cc:1033]:
```suggestion
  auto len_ist = static_cast<gdv_int32>(strlen("Asia/Kolkata"));
  auto len_pst = static_cast<gdv_int32>(strlen("America/Los_Angeles"));
```

File: cpp/src/gandiva/gdv_function_stubs_test.cc
Comment (kou) [inline cpp/src/gandiva/gdv_function_stubs_test.cc:1051]: How about adding error case for `from_utc_timezone_timestamp()` too?

Comment (palak-9202) [reply cpp/src/gandiva/gdv_function_stubs_test.cc:1067]: I've added all the suggested changes

**Review by kou — APPROVED** [review]
+1

Thanks. I've pushed some minor fixes.

I'll merge this once @anthonylouisbsb confirms error message generation code.

File: cpp/src/gandiva/gdv_function_stubs.cc
Comment (projjal) [inline cpp/src/gandiva/gdv_function_stubs.cc:628]: you can't use strlen since timezone is not a null terminated string. You can simply do the following
```
std::string err_msg = std::string(timezone, length) + " is an invalid time zone name.";
gdv_fn_context_set_error_msg(context, err_msg.c_str());
return 0;

**Review by projjal — APPROVED** [review]

**Pull request merged**
